cdc/sink/dmlsink/mq/transformer/columnselector/column_selector.go

```diff
@@ -0,0 +1,186 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package columnselector
+
+import (
+	filter "github.com/pingcap/tidb/util/table-filter"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/sink/dmlsink/mq/dispatcher"
+	"github.com/pingcap/tiflow/cdc/sink/dmlsink/mq/dispatcher/partition"
+	"github.com/pingcap/tiflow/pkg/config"
+	"github.com/pingcap/tiflow/pkg/errors"
+)
+
+type selector struct {
+	tableF  filter.Filter
+	columnM filter.ColumnFilter
+}
+
+func newSelector(
+	rule *config.ColumnSelector, caseSensitive bool,
+) (*selector, error) {
+	tableM, err := filter.Parse(rule.Matcher)
+	if err != nil {
+		return nil, errors.WrapError(errors.ErrFilterRuleInvalid, err, rule.Matcher)
+	}
+	if !caseSensitive {
+		tableM = filter.CaseInsensitive(tableM)
+	}
+	columnM, err := filter.ParseColumnFilter(rule.Columns)
+	if err != nil {
+		return nil, errors.WrapError(errors.ErrFilterRuleInvalid, err, rule.Columns)
+	}
+
+	return &selector{
+		tableF:  tableM,
+		columnM: columnM,
+	}, nil
+}
+
+// Match implements Transformer interface
+func (s *selector) Match(schema, table string) bool {
+	return s.tableF.MatchTable(schema, table)
+}
+
+// Apply implements Transformer interface
+// return error if the given event cannot match the selector, or the column cannot be filtered out.
+// the caller's should make sure the given event match the selector first before apply it.
+func (s *selector) Apply(event *model.RowChangedEvent) error {
+	// defensive check, this should not happen.
+	if !s.Match(event.Table.Schema, event.Table.Table) {
+		return errors.ErrColumnSelectorFailed.GenWithStack(
+			"the given event does not match the column selector, table: %v", event.Table)
+	}
+
+	for idx, column := range event.Columns {
+		if s.columnM.MatchColumn(column.Name) {
+			continue
+		}
+		if column.Flag.IsHandleKey() || column.Flag.IsUniqueKey() {
+			return errors.ErrColumnSelectorFailed.GenWithStack(
+				"primary key or unique key cannot be filtered out by the column selector, "+
+					"table: %v, column: %s", event.Table, column.Name)
+		}
+		event.Columns[idx] = nil
+	}
+
+	for idx, column := range event.PreColumns {
+		if s.columnM.MatchColumn(column.Name) {
+			continue
+		}
+		if column.Flag.IsHandleKey() || column.Flag.IsUniqueKey() {
+			return errors.ErrColumnSelectorFailed.GenWithStack(
+				"primary key or unique key cannot be filtered out by the column selector, "+
+					"table: %v, column: %s", event.Table, column.Name)
+		}
+		event.PreColumns[idx] = nil
+	}
+
+	return nil
+}
+
+// ColumnSelector manages an array of selectors, the first selector match the given
+// event is used to select out columns.
+type ColumnSelector struct {
+	selectors []*selector
```

Comment:
why does it need lots of selector?

Reply:
Each `selector` matches a subset of tables.